### DISCUSSION TITLE
settings_parser.y compilation

### DISCUSSION THREAD

**Discussion by yusuf-sami-acikgoz** [strongswan/strongswan#1596] [Q&A] [answered]
Hello, I'm trying to cross-compile the strongswan program for ARM architecture on two different linux-based machines (with different operating systems). On the first machine, the compilation process is successful, but on the second machine, the compilation process ends with the following error message. What could be the cause of the error? What should I check to fix the error. Thank you.

**Error message: settings_parser.y:72.20-37: syntax error, unexpected {...}**

**Comment by tobiasbrunner** [comment]
Make sure you have `bison` and `flex` installed (recent versions as well).

**Reply by yusuf-sami-acikgoz** [reply]
Thank you very much for your reply. 
When I check, I see that the bison and flex programs are installed.
I am able to compile strongswan-5.6.0 but I am getting error on 6.0.0beta3

**Reply by tobiasbrunner** [reply]
My guess is that the `bison` version you are using is too old. I think the syntax with braces for `api.prefix` was added with version 3.0 (which was released about 10 years ago).

**Reply by yusuf-sami-acikgoz** [reply]
After updating the correct bison version you said, the error I mentioned above was fixed. This time I get the following error:
**error: conflicting types for 'settings_parser_lex'**

When I check the flex version, I see that 2.6.4 is the most up-to-date version. Where should I check so I can fix this issue?

**Reply by tobiasbrunner** [reply]
What version of `bison` did you use exactly? (3.0 is certainly not the best choice as it's very old.) Also what are the conflicting definitions? (I guess it indicates that `YYSTYPE` is not `SETTINGS_PARSER_STYPE` for some reason. There should be a `#define` in the generated `settings_parser.c` file for the former and a definition for the latter in the generated `settings_parser.h` file.)

**Reply by yusuf-sami-acikgoz** [reply]
Thank you very much for your reply.
The bison version I'm using is 3.8

When I look at the settings_parser.c file, I see that **#define YYSTYPE SETTINGS_PARSER_STYPE** is defined. In the settings_parser.h file. In the settings parser.h file, I see that there is a union definition SETTINGS PARSER SKYPE between preprocessor macros, but I couldn't see anything like #define SETTINGS_PARSER_STYPE.

To explain better, I am sharing the error and note messages I received below:

settings/settings_parser.y:77:7: error: conflicting types for 'settings_parser_lex'
settings/settings_parser.y:78:2: note: in expansion of macro 'YY_DECL'
settings/settings_parser.h:113:5:note: previous declaration of 'settings_parser_lex' was here
settings/settings_parser.y:33.5: error: conflicting types for 'settings_parser_lex'
settings/settings_parser.h:113:5:note: previous declaration of 'settings_parser_lex' was here
settings/settings_parser.y:48:13: error: static decleration of 'settings_parser_error' follows non-static decleration
settings/settings_parser.h:110:6: note: previous declaration of 'settings_parser_error' was here

**Reply by tobiasbrunner** [reply]
Hm, I can't reproduce this with 3.8.2. The errors sound weird anyway. In particular the error regarding `settings_parser_error` as there should only be our own (static) definition. And I also don't see a definition of it in `settings_parser.h` generated here. Maybe your file was generated by the previous version? I'd try deleting the generated files (you could try `make maintainer-clean` which removes pretty much everything, then run `autogen.sh` and `configure` again).

**Reply by yusuf-sami-acikgoz** [reply]
Thank you for your answer. make maintainer-clean didn't work.
Many different linux tools are required for code compilation. What are the minimum versions of linux tools required to compile version 6.0.0beta3 ? (eg bison, flex, automake, autoconf etc.) I'm sorry if there is a list somewhere.

**Reply by tobiasbrunner** [reply]
> make maintainer-clean didn't work.

You mean the command didn't work at all, or that you still get the same error afterwards? If the latter, I really can't help you as I'm unable to reproduce the issue.

> Many different linux tools are required for code compilation. What are the minimum versions of linux tools required to compile version 6.0.0beta3 ?

Please note that if you [build from a release tarball](https://docs.strongswan.org/docs/5.9/install/install.html#_building_strongswan) you don't need any special tools (no bison, no flex, no gperf, no autotools). If you [build from the repository](https://docs.strongswan.org/docs/5.9/install/install.html#_building_from_the_git_repository), the list of required tools can be found in `HACKING`. However, I've no idea what the minimum versions for them are (but strongSwan built fine on e.g. Ubuntu 16.04 not too long ago, so they can't be too recent).

It's also always possible to build a release tarball on a current system (`make dist-bzip2`) and then use that to build on an older one.

**Comment by yusuf-sami-acikgoz** [comment, marked as answer]
> What version of `bison` did you use exactly? (3.0 is certainly not the best choice as it's very old.) Also what are the conflicting definitions? (I guess it indicates that `YYSTYPE` is not `SETTINGS_PARSER_STYPE` for some reason. There should be a `#define` in the generated `settings_parser.c` file for the former and a definition for the latter in the generated `settings_parser.h` file.)
